cmd/object-handlers-common.go

```diff
@@ -264,3 +266,69 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
 		}
 	}
 }
+
+// ilmLimitNoncurrentVersions limits the number of noncurrent versions of an
+// object based on lifecycle rules configured on the bucket.
+func ilmLimitNoncurrentVersions(ctx context.Context, objectAPI ObjectLayer, bucket, object string) (err error) {
+	var lc *lifecycle.Lifecycle
+	if lc, err = globalLifecycleSys.Get(bucket); err != nil {
+		if errors.Is(err, BucketLifecycleNotFound{Bucket: bucket}) {
+			return nil
+		}
+		return err
+	}
+
+	objOpts := lifecycle.ObjectOpts{Name: object}
+	lim := lc.NoncurrentVersionsExpirationLimit(objOpts)
+	if lim == 0 { // no limit set
+		return nil
+	}
+
+	var (
+		noncurrentVersions []ObjectToDelete
+		info               ListObjectVersionsInfo
+		keep               int
+	)
+	for {
+		info, err = objectAPI.ListObjectVersions(ctx, bucket, object, info.NextMarker, info.NextVersionIDMarker, "", 1000)
```

Comment:
No. This is way too expensive as an ILM call. You must find another way for this, otherwise you will tank performance. Callers should have enough information to decide if we need to truncate versions before they even make the call.

If you cannot decide this in the caller it is better NOT to have the call.

For scanner, we have all versions [here](https://github.com/minio/minio/blob/bb639d9f29fdd7595a1b520d8ea533c83acaa607/cmd/xl-storage.go#L458) - although only from one disk. Use that as a smoke screen test, and only do `ListObjectVersions` (or preferably a cheaper call) when you have seen more than the number of versions you want.

In terms of structure, add a second callback similar to `item.applyActions(...)`, that sends all versions and return a possibly modified list of versions.

Callback could be something like:
```Go
		fivs, err = item.applyVersionActions(ctx, fivs)
		if err != nil {
			if intDataUpdateTracker.debug {
				console.Debugf(color.Green("scannerBucket:")+" applying version actions failed: %v: %w\n", item.Path, err)
			}
			return sizeSummary{}, errSkipFile
		}

```

Reply:
This is only considered as faster solution @klauspost but not an optimal one.

Since we have to revisit this area subsequently once #13573

Reply:
@harshavardhana It will be *very* slow. We should not do a release with this IMO.

Reply:
As per @krisis testing, it isn't since its per object @klauspost - perhaps run `warp` and lets see how much we take a hit with and without it.

Reply:
@klauspost this function is called in `PutObjectHandler`, `CopyObjectHandler` and `CompleteMultipartUploadHandler`. It is not called by the scanner. It is only ever called for objects uploaded to a bucket with `MaxNoncurrentVersions` configured.
The warp tests I ran locally were to compare the performance difference between master (then) and this PR, *without* `MaxNoncurrentVersions` configured. This was to learn if this feature affects performance when disabled. I didn't find any noticeable difference.

Initially, I set out to implement this by adding a DeleteNoncurrentVersions method to the `ObjectLayer` and `StorageAPI` interface in order to push down the `MaxNoncurrentVersions` value. This way the listing/deletion happens at the participating disks. @harshavardhana recommended the current approach knowing that we will re-implement this taking advantage of https://github.com/minio/minio/pull/13573.

> If you cannot decide this in the caller it is better NOT to have the call.

With the background out of the way, I am open to the idea of moving application of this ILM rule exclusively to the scanner. At the scanner, we can do as you suggest. I believe that the scanner will be able to catch up on those objects whose versions are growing at a very fast rate, to remove excess noncurrent versions in batches. Does this sound good to you?